README.md
![CI status](https://codeship.com/projects/70b6f6b0-0892-0134-2fd9-0a16241ce4dd/status?branch=dev)

# Beit-Hatfusot Front-End README

Welcome to the frontend of the museum of the jewish people.  The code is in
Javascript using Angular 1.4.x and licensed under AGPLv3.  Please feel free to
fork and send us pull requests.

Our distant backend is the system known as BHP, an ancient Windows based
client-server software still used by our team of editors, volunteers and
museum visitors.  Our direct backend is an API server written in Python using
Flask, accessing data in a Mongo & ElasticSearch cluster.  The cluster is filled
with data from BHP using our migration bot.

## Blog Posts

- [I toggled public, now what?](https://medium.com/@alonisser/i-toggled-public-now-what-6b42959db251)
- [Freeing a Yack](https://medium.com/@daonb/freeing-a-yack-fc3799099eba)


## Dependencies

    $ sudo npm install -g gulp bower

## Getting the code and dependecies

    $ mkdir bh
    $ cd bh
    $ git clone http://....
    $ npm install
    $ bower install

## Run the live reloading development server

    $ gulp serve

Then head to `http://localhost:3000` in your browser.

## Test & Build

    $ gulp test
    $ gulp build-all

To send request to different server than the default `local` server
`http://localhost:5000`, set the `API_SERVER` environment variable to one of
`live`, `test`, `local`:

    $ API_SERVER=live gulp serve

## Distrubution

To build a distribution for live deployment, run:

    $ API_SERVER=live gulp dist

Don't forget to set `API_SERVER` to the desired one.

If you like to test the built version using the local development server, after
`gulp dist` run:

    $ gulp serve:dist

## Contributing

Contributions from both Jews and non-Jews are welcomed! We even have a
`beginner` label to help you start with (hopefully) simple issues.
Once you have an issue, just follow these simple steps:

1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Commit your changes: `git commit -am 'Add some feature'`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request :D

## Documentation

[Change Log](CHANGELOG.md)

We have very partial documentation so far. What we have, had been generated by
[generator-ngdoc](https://www.npmjs.com/package/generator-ngdoc)
After generation, the (generated) src folder have been removed, since it was
not needed.

In order to serve the docs http://localhost:3000 run:

    $ cd docs
    $ bower install
    $ gulp docs:serve


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased
### Changed
- Change "Fork me on Github" design
- Change meta-description for place items

### Fixed
- Persons Search query title
- Places Map: display red icons on top of blue ones
- Reset empty fields operators in Persons Search before producing new search 

### Added
- A User can add person to his story from Persons Search results page
- Indicate the error code in "Sorry, failed to fetch item" notification
- Meta tag to verify site ownership

## [0.10.0] - 2017-01-18
### Fixed
- Display persons' dates and places in persons search results
- Clean all the search results before starting a new search
- Family Trees: in case there are no years, don't show questions marks
- Make placeholder texts in the Person Search disappear once field has focus

### Added
- Add maps to places item page
- Return "Back to Search Results" button to family trees

## [0.9.14] - 2016-12-29
Give home page images meaningful names

## [0.9.13] - 2016-12-29
### Fixed

## [0.9.12] - 2016-12-22
### Fixed
- Update GA when navigating the tree
- Use Elastic Search for Wizard suggestions
- In wizard, exact match is case insensitive
- Fix persons date when navigating trees

## [0.9.11] - 2016-11-29
### Fixed
- Recentley viewed thumbnails are back
- Clear Europeana search result when sending a new search term
- Add feedback message for unchanged person search
- "Add to My story" button is no longer confused by items already in the story

## [0.9.10] - 2016-11-16
### Added
- Add Hebrew "Add to My Story" button 
- Display notification for items that only exist in one language
- A migration process now runs every 5 minutes, bringing new data from BHP
- Person URL now includes a tree version to better support updating of trees

### Fixed
- Fix Internet Explorer display issues
- On Item Page, display thumbnails for places
- Fix fetch more persons when query contains modificators

### Changed
- Redesign homepage
- Make MJS homepage for signed-in users
- Improve navigation in recently viewed
- Split "Images & Videos" tab in search ressults
- Related items are based only on BHP data

## [0.9.9] - 2016-09-07
### Added
- meta tags for item and my story pages
- caching for collections (in the API server)

### Fixed
- Consecutive searches - #161
- Clear the page title before navigation
- Fix username on public view of a User's MJS
- Tree drag and release - #76

### Changed
- Redesign subheader
- My Story: the entire line clickable when adding/removing an item to/from branch

### API Server Fixes
- `/login`: Validate `next` when creating a user,
    fixing cases where the login link was in `next` and the user is locked out
- `/user`: remove POST

## [0.9.8] - 2016-08-16
### Added
- A User can share his story
- A User can share another user's story

### Changed
- A User can have a name in both English and Hebrew languages
- A user name can be up to 20 chars long (12 before)
- A User can use numeric characters in the MJS tabs.
- Luminaries urls begins with `/luminary'

### Fixed
- Fix Internet Explorer display issues
- Family tree welcome dialog appears one per session

## [0.9.7] - 2016-07-11
### Added
- A User can add persons to his story

## [0.9.6] - 2016-06-29
### Added
- Display Photo Unit Number

### Changed
- Using `gulp` instead of `grunt` for devOps
- Using GTM instead of analytics
- Github ribbon is now redo
- Recently Viewed now displays all the items than it stores
- Typos
- User rename now works as expected (was flaky)
- Limiting branch name in my story to 26 chars
- Unlimit the number of item in my story
- Improved page titles

## [0.9.5] - 2016-06-14
### Added
- A Github ribbon
- use gravatar for avatars
- `grunt build-dist` to create a better distribution under `dist/`

### Changed
- '/mjs' is unaccesible for anonymous users
- smoother item remove in mjs
- fix persons search form when switching from "more fields" to "less fields"
- improve google analytics integration so that each state change is recorded
- clear "loading" gif when entering a state
- use unicode's unknown char when missing a person's date
- improve grunt portability (fixing issues with codeship)
- `fab` is now back to deploying on our server, instead of compute storage which
returns 404 for any specific address other then "index.html"
- support up to 50 items in mjs


gulpfile.js
'use strict';

var gulp = require('gulp'),
    g = require('gulp-load-plugins')({lazy: false}),
    noop = g.util.noop,
    es = require('event-stream'),
    bowerFiles = require('main-bower-files'),
    rimraf = require('rimraf'),
    queue = require('streamqueue'),
    lazypipe = require('lazypipe'),
    bower = require('./bower'),
    runSequence = require('run-sequence'),
    isWatching = false;

var htmlminOpts = {
  removeComments: true,
  /*collapseWhitespace: true,*/
  removeEmptyAttributes: false,
  collapseBooleanAttributes: true
};

/**
 * Base url in config.js
 */
var _baseUrlEnvMapping = {
	local: "localhost:5000",
	test: "devapi.dbs.bh.org.il",
	live: "api.dbs.bh.org.il"
};

gulp.task('base-url', function() {
  var apiServer = process.env.API_SERVER || 'test';

  return gulp.src('./js/modules/config/config.js')
    .pipe(g.replace('BaseUrlPlaceHolder', _baseUrlEnvMapping[apiServer]))
    .pipe(gulp.dest('./.tmp/'));
});

/**
 * CSS
 */
gulp.task('clean-css', function (done) {
  rimraf('./.tmp/css', done);
});

gulp.task('styles', ['clean-css'], function () {
  return gulp.src([
    './scss/**/*.scss',
    '!./scss/**/_*.scss'
  ])
    .pipe(g.sourcemaps.init())
    .pipe(g.sass().on('error', g.sass.logError))
    .pipe(g.autoprefixer({browsers: ['last 2 versions', '> 2%']}))
    .pipe(g.sourcemaps.write('./'))
    .pipe(gulp.dest('./.tmp/css/'))
    .pipe(g.cached('built-css'))
    .pipe(livereload());
});

gulp.task('styles-dist', ['styles'], function () {
  return cssFiles().pipe(dist('css', bower.name));
});

/**
 * Scripts
 */
gulp.task('scripts-dist', ['templates-dist'], function () {
  return appFiles().pipe(dist('js', bower.name, {ngAnnotate: true}));
});

/**
 * Templates
 */
gulp.task('templates', function () {
  return templateFiles().pipe(buildTemplates());
});

gulp.task('templates-dist', function () {
  return templateFiles({min: true}).pipe(buildTemplates());
});

/**
 * Vendors
 */
gulp.task('vendors', function () {
  var bowerStream = gulp.src(bowerFiles());
  return es.merge(
    bowerStream.pipe(g.filter('**/*.css')).pipe(dist('css', 'vendors')),
    bowerStream.pipe(g.filter('**/*.js')).pipe(dist('js', 'vendors'))
  );
});

/**
 * Index
 */
gulp.task('index', index);
gulp.task('build-all', ['styles', 'templates', 'base-url'], index);

function index () {
  var opt = {read: false};
  return gulp.src('./index.html')
    .pipe(g.inject(gulp.src(bowerFiles(), opt), {ignorePath: 'bower_components', starttag: '<!-- inject:vendor:{{ext}} -->'}))
    .pipe(g.inject(es.merge(appFiles(), cssFiles(opt)), {ignorePath: ['.tmp']}))
    .pipe(gulp.dest('.'))
    .pipe(g.embedlr())
    .pipe(gulp.dest('./.tmp/'))
    .pipe(livereload());
}



/**
 * Static assets
 */

gulp.task('fonts', function () {
  return gulp.src('./fonts/**')
    .pipe(gulp.dest('./dist/fonts'));
});

gulp.task('images', function () {
  return gulp.src('./images/**')
    .pipe(gulp.dest('./dist/images'));
});

gulp.task('assets', ['fonts', 'images']);


/**
 * Dist
 */
gulp.task('dist-clean', function (done) {
  rimraf('./dist', done);
});

gulp.task('dist-index', function() {
  return gulp.src('./index.html')
    .pipe(g.inject(gulp.src('./dist/{js,css}/vendors*.min.{js,css}'), {ignorePath: 'dist', starttag: '<!-- inject:vendor:{{ext}} -->', addRootSlash:false}))
    .pipe(g.inject(gulp.src('./dist/{js,css}/' + bower.name + '*.min.{js,css}'), {ignorePath: 'dist', addRootSlash:false}))
    .pipe(g.htmlmin(htmlminOpts))
    .pipe(gulp.dest('./dist/'));
});

gulp.task('dist', function (cb) {
  runSequence(
    'dist-clean',
    ['base-url', 'vendors', 'assets', 'styles-dist', 'scripts-dist'],
    'dist-index',
    cb
  );
});


/**
 * Static file server
 */
gulp.task('connect', function() {

  g.connect.server({
    port: 3000,
    root: ['./.tmp', './', './bower_components'],
    fallback: './index.html',
    middleware: function() {
      return [
        // place required middleware here
      ];
    }
  });
});

gulp.task('serve:dist', function() {

  g.connect.server({
    port: 3000,
    root: ['./dist'],
    fallback: './dist/index.html',
    middleware: function() {
      return [
        // place required middleware here
      ];
    }
  });
});

/**
 * Watch
 */
gulp.task('serve', ['watch']);
gulp.task('watch', ['connect', 'default'], function () {
  isWatching = true;
  // Initiate livereload server:
  g.livereload.listen();
  gulp.watch('./js/**/*.js').on('change', function (evt) {
    if (evt.type !== 'changed') {
      gulp.start('index');
    } else {
      g.livereload.changed(evt);
    }
  });
  gulp.watch('./index.html', ['index']);
  gulp.watch(['./templates/**/*.html'], ['templates']);
  gulp.watch(['./scss/**/*.scss']).on('change', function (evt) {
    if (evt.type !== 'changed') {
      gulp.start('index');
    } else {
      gulp.start('styles');
    }
  });
});

/**
 * Default task
 */
gulp.task('default', ['build-all']);


/**
 * Test
 */
gulp.task('test', ['templates', 'base-url'], function () {
  return testFiles()
    .pipe(g.karma({
      configFile: 'js/test/karma.conf.js',
      action: 'run'
    }));
});

/**
 * Inject all files for tests into karma.conf.js
 * to be able to run `karma` without gulp.
 */
gulp.task('karma-conf', ['templates', 'base-url'], function () {
  return gulp.src('./js/test/karma.conf.js')
    .pipe(g.inject(testFiles(), {
      starttag: 'files: [',
      endtag: ']',
      addRootSlash: false,
      transform: function (filepath, file, i, length) {
        return '  \'' + filepath + '\'' + (i + 1 < length ? ',' : '');
      }
    }))
    .pipe(gulp.dest('./js/test/'));
});

/**
 * Test files
 */
function testFiles() {
  return new queue({objectMode: true})
    .queue(gulp.src('./bower_components/jquery/dist/jquery.js'))
    .queue(gulp.src('./bower_components/angular/angular.js'))
    .queue(gulp.src(bowerFiles()).pipe(g.filter('**/*.js')))
    .queue(gulp.src('./bower_components/angular-mocks/angular-mocks.js'))
    .queue(appFiles())
    .queue(gulp.src(['./js/**/*Spec.js', './.tmp/js/**/*Spec.js']))
    .done();
}

/**
 * All CSS files as a stream
 */
function cssFiles (opt) {
  return gulp.src('./.tmp/css/**/*.css', opt);
}

/**
 * All AngularJS application files as a stream
 */
function appFiles () {
  var files = [
    './.tmp/' + bower.name + '-templates.js',
    './.tmp/config.js',
    './js/**/*.js',
    '!./js/**/*Spec.js',
    '!./js/modules/config/**',
    '!./js/test/**',
    '!./js/modules/*/test/**'
  ];
  return gulp.src(files)
    .pipe(g.naturalSort())
    .pipe(g.angularFilesort());
}

/**
 * All AngularJS templates/partials as a stream
 */
function templateFiles (opt) {
  return gulp.src(['./templates/**/*.html'], opt)
    .pipe(opt && opt.min ? g.htmlmin(htmlminOpts) : noop());
}

/**
 * Build AngularJS templates/partials
 */
function buildTemplates () {
  return lazypipe()
    .pipe(g.ngHtml2js, {
      moduleName: bower.name + '-templates',
      /*prefix: '/' + bower.name + '/',*/
      prefix: 'templates/'
    })
    .pipe(g.concat, bower.name + '-templates.js')
    .pipe(gulp.dest, './.tmp')
    .pipe(livereload)();
}

/**
 * Concat, rename, minify
 *
 * @param {String} ext
 * @param {String} name
 * @param {Object} opt
 */
function dist (ext, name, opt) {
  opt = opt || {};
  var stamp = new Date().getTime();
  return lazypipe()
    .pipe(g.concat, name + '.' + ext)
    // .pipe(gulp.dest, './dist/' + ext + '/')
    // .pipe(opt.ngAnnotate ? g.ngAnnotate : noop)
    // .pipe(opt.ngAnnotate ? g.rename : noop, name + '.annotated.' + ext)
    // .pipe(opt.ngAnnotate ? gulp.dest : noop, './dist/' + ext)
    .pipe(ext === 'js' ? g.uglify : g.minifyCss)
    .pipe(g.rename, name + '-' + stamp + '.min.' + ext)
    .pipe(gulp.dest, './dist/' + ext + '/')();
}

/**
 * Livereload (or noop if not run by watch)
 */
function livereload () {
  return lazypipe()
    .pipe(isWatching ? g.livereload : noop)();
}
